internal/terraform/node_resource_abstract_instance.go

```diff
@@ -1594,10 +1594,23 @@ func (n *NodeAbstractResourceInstance) readDataSource(ctx EvalContext, configVal
 		}
 	} else {
 		resp = provider.ReadDataSource(providers.ReadDataSourceRequest{
-			TypeName:     n.Addr.ContainingResource().Resource.Type,
-			Config:       configVal,
-			ProviderMeta: metaConfigVal,
+			TypeName:        n.Addr.ContainingResource().Resource.Type,
+			Config:          configVal,
+			ProviderMeta:    metaConfigVal,
+			DeferralAllowed: ctx.Deferrals().DeferralAllowed(),
 		})
+
+		if resp.Deferred != nil {
+			deffered := ctx.Deferrals()
+			deffered.ReportResourceInstanceDeferred(n.Addr, resp.Deferred.Reason, &plans.ResourceInstanceChange{
```

Comment:
I think there's a dedicated ReportDataSourceDeferred function? We don't need to save the whole change for the data source in the Deferred object and the alternate function should have a different signature.

Reply:
I looked into it and there is a `ReportDataSourceExpansionDeferred` but in this case the expansion is not deferred but the instance itself. I am a bit confused in general why we sometimes distinguish between managed and not managed resources and sometimes resources and data sources. I though `ReportResourceInstanceDeferred` refers to both managed and unmanaged resources.

Reply:
So a managed resource is the same as a resource in the config and unmanaged resource is the same as a data source in the config. Unmanaged resource and data source mean the same thing.

For data sources, we don't actually write their outputs into the plan at all. Everything is just recomputed during the apply stage for data sources (I think). This means we don't need to create a change for them as we will never refer to it.

I think we should create another function for `ReportDataSourceInstanceDeferred` within the deferred object. And it should behave the same way as the difference between the instance deferred and expansion deferred for the resources do, but use an `addrs.Set` with the addresses instead of the `addrs.Map`.

Reply:
Ok yeah, that makes sense to me 👍